subprojects/gradle/runner/client/src/main/kotlin/com/avito/runner/scheduler/metrics/TestMetricsAggregatorImpl.kt

```diff
@@ -0,0 +1,59 @@
+package com.avito.runner.scheduler.metrics
+
+import com.avito.runner.scheduler.metrics.model.TestKey
+import com.avito.runner.scheduler.metrics.model.TestTimestamps
+import org.nield.kotlinstatistics.median
+
+internal class TestMetricsAggregatorImpl(
+    private val testSuiteStartedTime: Long,
+    private val testSuiteEndedTime: Long,
+    testTimestamps: Map<TestKey, TestTimestamps>
+) : TestMetricsAggregator {
+
+    private val firstTestStarted = testTimestamps.values
+        .mapNotNull { it.started }
+        .minOrNull()
+
+    private val lastTestEnded = testTimestamps.values
+        .mapNotNull { it.finished }
+        .maxOrNull()
+
+    private val queueTimes: List<Long> = testTimestamps.values
+        .mapNotNull { it.onDevice }
+        .map { it - testSuiteStartedTime }
+
+    private val installationTimes: List<Long> = testTimestamps.values
```

Comment:
Storing filtered values seems like premature optimization.
They are needed only one time and can be easily calculated in place.

Reply:
i can wrap it in `lazy {}`, but i wanted to extract this calculation into functions anyways, and why do it as a function if it's value doesn't change?

Reply:
Of course, in this class with tests, we shouldn't see a difference. I'm talking about the general approach.

The more state we have, the more complexity we add.
Storing something is needed only for caching when a computation is long. It is in fact a performance technique.
We consume more memory for faster access to results.
I prefer to see caching when we have reasons to think that calculation in place / by request is not enough.

Reply:
i see, i believe it make sense here, but i agree in general